crates/wasi-provider/src/states/initializing.rs

```diff
@@ -0,0 +1,138 @@
+use std::collections::HashMap;
+
+use log::{error, info};
+
+use crate::PodState;
+use k8s_openapi::api::core::v1::Pod as KubePod;
+use kube::api::{Api, PatchParams};
+use kubelet::container::{ContainerKey, Status as ContainerStatus};
+use kubelet::pod::{key_from_pod, Handle};
+use kubelet::state::prelude::*;
+
+use super::error::Error;
+use super::starting::{start_container, ContainerHandleMap, Starting};
+
+async fn patch_init_status(
+    client: &Api<KubePod>,
+    pod_name: &str,
+    name: String,
+    status: &ContainerStatus,
+) -> anyhow::Result<()> {
+    // We need to fetch the current status because there is no way to merge with a strategic merge patch ere
+    let mut init_container_statuses = match client.get(pod_name).await {
+        Ok(p) => match p.status {
+            Some(s) => s.init_container_statuses.unwrap_or_default(),
+            None => {
+                return Err(anyhow::anyhow!(
+                    "Pod is missing status information. This should not occur"
+                ));
+            }
+        },
+        Err(e) => {
+            error!("Unable to fetch current status of pod {}, aborting status patch (will be retried on next status update): {:?}", pod_name, e);
+            // FIXME: This is kinda...ugly. But we can't just
+            // randomly abort the whole process due to an error
+            // fetching the current status. We should probably have
+            // some sort of retry mechanism, but that is another
+            // task for another day
+            Vec::default()
+        }
+    };
+    match init_container_statuses.iter().position(|s| s.name == name) {
+        Some(i) => {
+            init_container_statuses[i] = status.to_kubernetes(name);
+        }
+        None => {
+            init_container_statuses.push(status.to_kubernetes(name));
+        }
+    };
+    let s = serde_json::json!({
+        "metadata": {
+            "resourceVersion": "",
+        },
+        "status": {
+            "initContainerStatuses": init_container_statuses,
+        }
+    });
+    client
+        .patch_status(pod_name, &PatchParams::default(), serde_json::to_vec(&s)?)
+        .await?;
+    Ok(())
+}
+
+state!(
+    /// The Kubelet is running the init containers for the pod.
+    Initializing,
+    PodState,
+    {
+        let client: Api<KubePod> = Api::namespaced(
```

Comment:
This all looks fine in itself, but is strong evidence for the macro rework.  This is a 70-line function of some complexity; maintaining it would be a lot easier if the macro allowed for better editor support.

Reply:
yep